src/formatDate.js

```diff
@@ -50,7 +50,81 @@
  */
 
 function formatDate(date, fromFormat, toFormat) {
-  // write code here
+  const inputSeparator = fromFormat[fromFormat.length - 1];
+  const inputDateComponents = date.split(inputSeparator);
+
+  const outputSeparator = toFormat[toFormat.length - 1];
+  const outputDateComponents = [];
+
+  const dateComponents = {
+    day: '',
+    month: '',
+    year: '',
+    yearFormat: '',
+    dateFactory() {
+      for (let i = 0; i < fromFormat.length - 1; i++) {
+        switch (fromFormat[i]) {
+          case 'DD':
+            this.day = inputDateComponents[i];
+            break;
+          case 'MM':
+            this.month = inputDateComponents[i];
+            break;
+          case 'YY':
+          case 'YYYY':
+            this.year = inputDateComponents[i];
+            this.yearFormat = fromFormat[i];
+            break;
+          default:
+            console.error(`Wrong format: ${fromFormat[i]}`);
+        }
+      }
+    },
+    formatFactory() {
+      for (let i = 0; i < toFormat.length - 1; i++) {
+        const currentFormat = toFormat[i];
+
+        switch (currentFormat) {
+          case 'DD':
+            outputDateComponents.push(this.day);
+            break;
+          case 'MM':
+            outputDateComponents.push(this.month);
+            break;
+          case 'YY':
+          case 'YYYY':
+            const outputYear = this.convertYearFormat(
+              this.year,
+              this.yearFormat,
+              currentFormat
+            );
+
+            outputDateComponents.push(outputYear);
+            break;
+          default:
+            console.error(`Wrong format: ${currentFormat}`);
+        }
+      }
+    },
+    convertYearFormat(year, previousFormat, targetFormat) {
+      if (previousFormat === targetFormat) {
+        return year;
+      }
+
+      if (targetFormat === 'YY') {
+        return year.slice(-2);
+      }
+
+      const prefix = year < 30 ? '20' : '19';
+
+      return prefix + year;
+    },
+  };
+
+  dateComponents.dateFactory();
+  dateComponents.formatFactory();
```

Comment:
for simplicity it's may be better to define the functions as standalone functions outside of an object. In this case for dateFactory and formatFactory you need to reference the object dateComponents every time you want to call a function, inside of it. It wouldn't work if you moved it outside of the formatDate function. convertYearFormat is an example of a function with a good purpose that actually helps and might be reused. However, move it outside of the formatDate function. date and format factories functions are not bringing any value, you can remove them and just run the code inside of this function.